Vault/Library/CachePersistence.cs

```diff
@@ -124,10 +128,93 @@ void AfterAccessNotification(AuthenticationRecord authRecord)
                 // Call AuthenticateAsync to fetch a new AuthenticationRecord.
                 authRecord = credential.Authenticate();
 
+                // Create the original data to be encrypted
+                var toEncrypt = Encoding.ASCII.GetBytes(authRecord.ToString());
+
+                // Create some random entropy.
+                byte[] entropy = CreateRandomEntropy();
+
                 // Serialize the AuthenticationRecord to disk so that it can be re-used across executions of this initialization code.
-                var authRecordStream = new FileStream(FileName, FileMode.Create, FileAccess.Write);
-                authRecord.SerializeAsync(authRecordStream);
+                var authRecordStream = new FileStream(FileName, FileMode.OpenOrCreate);
+
+                // Encrypt a copy of the data to the stream.
+                int bytesWritten = EncryptDataToStream(toEncrypt, entropy, DataProtectionScope.CurrentUser, authRecordStream);
+
+                authRecordStream.Close();
             }
         }
+
+        public static byte[] CreateRandomEntropy()
+        {
+            // Create a byte array to hold the random value.
+            byte[] entropy = new byte[16];
+
+            // Create a new instance of the RNGCryptoServiceProvider.
+            // Fill the array with a random value.
+            new RNGCryptoServiceProvider().GetBytes(entropy);
+
+            // Return the array.
+            return entropy;
+        }
+
+        public static int EncryptDataToStream(byte[] Buffer, byte[] Entropy, DataProtectionScope Scope, Stream S)
+        {
+            if (Buffer == null)
+                throw new ArgumentNullException("Buffer");
+            if (Buffer.Length <= 0)
+                throw new ArgumentException("Buffer");
+            if (Entropy == null)
+                throw new ArgumentNullException("Entropy");
+            if (Entropy.Length <= 0)
+                throw new ArgumentException("Entropy");
+            if (S == null)
+                throw new ArgumentNullException("S");
+
+            int length = 0;
+
+            // Encrypt the data and store the result in a new byte array. The original data remains unchanged.
+            byte[] encryptedData = ProtectedData.Protect(Buffer, Entropy, Scope);
+
+            // Write the encrypted data to a stream.
+            if (S.CanWrite && encryptedData != null)
+            {
+                S.Write(encryptedData, 0, encryptedData.Length);
+
+                length = encryptedData.Length;
+            }
+
+            // Return the length that was written to the stream.
+            return length;
+        }
+
+        public static byte[] DecryptDataFromStream(byte[] Entropy, DataProtectionScope Scope, Stream S, int Length)

```

Comment:
I don't see where this is getting called.

Reply:
When decrypting data, we need an `AuthenticationRecord` parameter to implement silent authentication.
```
// Load the previously serialized AuthenticationRecord from disk and deserialize it.
var authRecordStream = new FileStream(FileName, FileMode.Open, FileAccess.Read);
authRecord = AuthenticationRecord.Deserialize(authRecordStream);

// Construct a new client with our TokenCachePersistenceOptions with the addition of the AuthenticationRecord property.
// This tells the credential to use the same token cache in addition to which account to try and fetch from cache when GetToken is called.
var credential = new InteractiveBrowserCredential(
    new InteractiveBrowserCredentialOptions
    {
        TokenCachePersistenceOptions = new TokenCachePersistenceOptions { Name = TOKEN_CACHE_NAME },
        AuthenticationRecord = authRecord
    });
```

But `ProtectedData.Unprotect(inBuffer, Entropy, Scope);` method return an `byte[] array` value. So we didn't call here, considering the possible updates after that, we kept it for the time being, do you have any ideas or suggestions?

Reply:
If you're not decrypting your cache, I don't see how it's useful. You could instead initially a `MemoryStream` from a `byte[]` array which will give you the stream you need.